tools/parsetrace.py

```diff
@@ -0,0 +1,197 @@
+#!python3
+############################################################################
+# tools/parsetrace.py
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.  The
+# ASF licenses this file to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+import argparse
+import bisect
+import os
+import re
+from typing import Union
+
+import cxxfilt
+import parse
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import SymbolTableSection
+from pydantic import BaseModel
+
+
+class SymbolTables(object):
+    def __init__(self, file):
+        elffile = open(file, "rb")
+        self.elffile = ELFFile(elffile)
+        self.symbol_dict = dict()
+        self.addr_list = list()
+
+    def __symbol_filter(self, symbol):
+        if symbol["st_info"]["type"] != "STT_FUNC":
+            return None
+        if symbol["st_info"]["bind"] == "STB_WEAK":
+            return None
+        if symbol["st_shndx"] == "SHN_UNDEF":
+            return None
+        return symbol
+
+    def __get_symtable(self):
+        symbol_tables = [
+            s
+            for _, s in enumerate(self.elffile.iter_sections())
+            if isinstance(s, SymbolTableSection)
+            and s.name == ".symtab"
+            and s["sh_entsize"]
+        ]
+
+        if not symbol_tables:
+            return None
+
+        return symbol_tables[0]
+
+    def parse_symbol(self):
+        if self.elffile is None:
+            return
+
+        symtable = self.__get_symtable()
+        for nsym, symbol in enumerate(symtable.iter_symbols()):
+            if self.__symbol_filter(symbol) is not None:
+                symbol_name = cxxfilt.demangle(symbol.name)
+                func_name = re.sub(r"\(.*$", "", symbol_name)
+                if func_name[0] == "_" or func_name.find(".") != -1:
+                    continue
+
+                self.symbol_dict[symbol["st_value"] & ~0x01] = func_name
+        self.addr_list = sorted(self.symbol_dict, reverse=False)
+
+    def addr2symbol(self, addr: int):
+        index = bisect.bisect(self.addr_list, addr) - 1
+        if index != -1:
+            return self.symbol_dict[self.addr_list[index]]
+        return "<%#x>: unknow function" % addr
```

Comment:
should we keep the original address without change

Reply:
The address is not changed here, it is the result of the binary search minus one
Used to find the symbol name less than or equal to the specified address